api/renter.go

```diff
@@ -350,49 +387,156 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
-	if err != nil {
-		WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
-		return
+	if p.Async { // Create goroutine if `async` param set.
+		go api.renter.DownloadSection(p)
+	} else {
+		err := api.renter.DownloadSection(p)
+		if err != nil {
+			WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
+			return
+		}
 	}
 
-	WriteSuccess(w)
+	if !p.Httpresp {
+		WriteSuccess(w)
+	}
 }
 
 // renterDownloadAsyncHandler handles the API call to download a file asynchronously.
 func (api *API) renterDownloadAsyncHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	// Ensure that the async flag is not set to false.
+	if strings.Contains(req.URL.RawQuery, "async=false") {
+		WriteError(w, Error{"/downloadasync does not support allow requests containing `async=false`."}, http.StatusBadRequest)
 		return
 	}
 
-	siapath := strings.TrimPrefix(ps.ByName("siapath"), "/")
-	fileExists := false
+	// Set async flag to true.
+	req.URL.RawQuery += "&async=true"
+	p, errmsg := api.parseAndValidateDownloadParameters(w, req, ps)
+	if errmsg != nil {
+		WriteError(w, *errmsg, http.StatusBadRequest)
+		return
+	}
+
+	go api.renter.DownloadSection(p)
+
+	WriteSuccess(w)
+}
+
+func (api *API) parseAndValidateDownloadParameters(w http.ResponseWriter, req *http.Request, ps httprouter.Params) (*modules.RenterDownloadParameters, *Error) {
+	destination := req.FormValue("destination")
+
+	// The offset and length in bytes.
+	offsetparam := req.FormValue("offset")
+	lengthparam := req.FormValue("length")
+
+	// Determines whether the response is written to response body.
+	httprespparam := req.FormValue("httpresp")
 
-	files := api.renter.FileList()
-	for _, file := range files {
-		if file.SiaPath == siapath {
-			fileExists = true
+	// Determines whether to return on completion of download or straight away.
+	// If httprespparam is present, this parameter is ignored.
+	asyncparam := req.FormValue("async")
+
+	// Parse the offset and length parameters.
+	var offset, length uint64
+	var err error
+	if len(offsetparam) > 0 {
+		offset, err = strconv.ParseUint(offsetparam, 10, 64)
+		if err != nil {
+			return nil, &Error{"could not decode the offset as uint64: " +
+				err.Error()}
 		}
 	}
+	if len(lengthparam) > 0 {
+		length, err = strconv.ParseUint(lengthparam, 10, 64)
+		if err != nil {
+			return nil, &Error{"could not decode the length as uint64: " +
+				err.Error()}
+		}
+	}
+	// Verify that if either offset or length have been provided that both were provided.
+	offparampassed := len(offsetparam) > 0
+	lenparampassed := len(lengthparam) > 0
+	if (offparampassed || lenparampassed) &&
+		!(offparampassed && lenparampassed) {
```

Comment:
this if statement can definitely be one line. We generally try not to wrap code because it tends to reduce readability. There are of course exceptions, but I think that's generally not until you are approaching 150 chars.